src/Core/Managed/Shared/DataContracts/TelemetryContext.cs

```diff
@@ -40,6 +41,7 @@ internal TelemetryContext(IDictionary<string, string> properties)
             Debug.Assert(properties != null, "properties");
             this.properties = properties;
             this.tags = new ConcurrentDictionary<string, string>();
+            this.correlationContext = new ConcurrentDictionary<string, string>();
```

Comment:
```ConcurrentDictionary<K,V>``` is pretty heavyweight. To be creating one on every telemetry item could be a performance problem. (I notice we already have one for ```tags```).
I have a few pieces of advice:
1. If the correlationContext is often empty, then use a lazy initializer to create the dictionary on first use. Take care that just _reading_ the empty context doesn't create a new dictionary.
2. Consider an ```ImmutableDictionary<K,V>``` (from System.Collections.Immutable).
3. If you must use a ```ConcurrentDictionary``` call the constructor that allows you to set the maximum concurrency, and set it to 1. (You also have to set the expected capacity, so I would set it to zero). That way you get thread-safety without the bulk of the overhead.